### PR TITLE
Text Encoder LR Scale for Stable Diffusion Training

### PR DESCRIPTION
**Pull request by sangoi-exe**
Hello! This is a suggested implementation to scale the magnitude of the Text Encoder's learning rate during Stable Diffusion training sessions.

So, I'm another user of the stable diffusion community, and, like many others, I really enjoy training using Prodigy. However, as we know, many people come here to find a solution to the overtraining of the TE, since Prodigy doesn't accept different LRs.

Just so you know, I've been using [Kohya's](https://github.com/kohya-ss/sd-scripts) scripts and training LoRAs for SD 1.5.

After analyzing how Prodigy receives and handles tensors, I came up with a trick to reduce the LR of the TE during the `.step()`.

Btw, correct me if I'm wrong, please.

I based this on the premise that TE tensors will always be 2D, so their shape will always have two dimensions.

As the tensors from Kohya come in two groups, using `all(len(p.shape) == 2 for p in group["params"])` allowed me to determine which group of tensors belongs to the TE.

With that information, I created a boolean and used an if-else statement for the weight update:

```
            is_text_encoder = all(len(p.shape) == 2 for p in group["params"])

            ...

                ### Take step
                # and here we apply the new te scale to the weight update
                if is_text_encoder:
                    # scale down the weight update for TE due his higher sensitivity
                    p.data.addcdiv_(exp_avg, denom, value=-(dlr * te_scale))
                else:
                    # standard weight update for other parameters
                    p.data.addcdiv_(exp_avg, denom, value=-dlr)
```

My approach will only influence the optimizer if the user sets the `te_scale` in the optimizer's extra arguments before starting the training. Otherwise, `te_scale` will be 1, and nothing will change in the weight update.

### CODE REVIEW

**Comment by konstmish** [comment]
Hi Lucas, thanks for preparing this pull request to improve the performance on SD. I like the idea of supporting different scales for different parts of a network, but ideally we'd want Prodigy to have parameters that make sense for all applications, not just SD. For other networks that may make this hyperparameter a distraction. Unfortunately, I cannot come up with a solution that would be easy to use from kohya-ss/sd-scripts, it seems that your solution makes a lot of sense in that context, though it's not related to the original purpose of this repository.
@adefazio, do you have any thoughts on this?

**Comment by DevArqSangoi** [comment]
@konstmish hello! Thanks a lot for the response! I totally get it, and even if it can’t be officially implemented, the people in charge of the SD trainers' repositories can look into finding a way to add it in a custom way, if they think it could be useful. 😊

**Comment by Panchovix** [comment]
Hey, sorry to bump an older PR, but would this still be applied to make it work on sd-scripts for SDXL? I'm having good results with Prodigy but the TE is def overtrained, and this would help some of the issues.

EDIT: I applied it now and see how it behaves.
